File: sorting/bst.py
class BST:
	def __init__(self, list=[]):
		self.root = None

		for value in list:
			self.insert(value)

	def insert(self, key):
		new_node = Node(key)

		if self.root is None:
			self.root = new_node
			return

		current = self.root

		while True:
			if key < current.key:
				if current.left is None:
					current.left = new_node
					new_node.parent = current
					break
				else:
					current = current.left
			else:
				if current.right is None:
					current.right = new_node
					new_node.parent = current
					break
				else:
					current = current.right

	def find(self, key):
		current = self.root

		while current is not None:
			if current.key == key:
				return current
			elif key < current.key:
				current = current.left
			else:
				current = current.right

		# Return None if key could not be found
		return None

	# Deletes the minimum value in the BST and returns it
	def delete_min(self):
		"""Delete the minimum value in the BST and return it."""
		if self.root is None:
			return None

		# Navigate to leftmost node
		current = self.root
		while current.left:
			current = current.left

		if current.parent:
			current.parent.left = current.right
		else: # If current node is root
			self.root = current.right

		if current.right:
			current.right.parent = current.parent

		# Remove min node
		current.disconnect()
		# Return min value
		return current.key

	def sort(self):
		"""Return a sorted list of the BST's keys."""
		self.list = []
		# Traverse the BST using DFS, appending keys along the way 
		self.traverse(self.root)
		return self.list

	def traverse(self, node):
		if node is None:
			return

		self.traverse(node.left)
		self.list.append(node.key)
		self.traverse(node.right)

class Node:
	def __init__(self, key):
		self.key = key
		self.disconnect()

	def disconnect(self):
		self.parent = None
		self.left = None
		self.right = None

length = input('Enter the number of values you wish to sort: ')
unsorted = []
for i in range(length):
	unsorted.append(input('Number %s: ' % (i + 1)))

bst = BST(unsorted)
sorted = bst.sort()
print sorted

File: sorting/bst.cpp
#include <iostream>
using namespace std;

// Global variable for traversing BST and modifying array
int i;

struct node {
	int key;
	node* parent;
	node* left;
	node* right;
};

void insert(node* root, int key) {
	if(key < root->key) {
		if(root->left == nullptr) {
			root->left = new node;
			root->left->key = key;
			root->left->parent = root;
			root->left->left = nullptr;
			root->left->right = nullptr;
		}
		else
			insert(root->left, key);
	}
	else {
		if(root->right == nullptr) {
			root->right = new node;
			root->right->key = key;
			root->right->parent = root;
			root->right->left = nullptr;
			root->right->right = nullptr;
		}
		else
			insert(root->right, key);
	}
}

node* find(node* root, int key) {
	if(key < root->key) {
		if(root->left == nullptr)
			return nullptr;
		else if(root->left->key == key)
			return root->left;
		else
			return find(root->left, key);
	}
	else {
		if(root->right == nullptr)
			return nullptr;
		else if(root->right->key == key)
			return root->right;
		else
			return find(root->right, key);
	}
}

node* find_min(node* root) {
	node* min = root;
	while(min->left != nullptr)
		min = min->left;
	return min;
}

// Find minimum value in BST and delete that node
// root parameter is a pointer passed by reference because we may modify the root node
int delete_min(node*& root) {
	node* min = find_min(root);

	if(min->parent != nullptr)
		min->parent->left = min->right;
	else // In the case that min node is root
		root = min->right;

	if(min->right != nullptr)
		min->right->parent = min->parent;

	return min->key;
}

node* next_larger(node* current) {
	if(current->right != nullptr)
		return find_min(current->right);
	else {
		node* y = current->parent;
		while(y != nullptr && current == y->right) {
			current = y;
			y = y->parent;
		}
		return y;
	}
}

// Builds a binary search tree from an array and returns the root node
node* build_binary_tree(int* list, int length) {
	if(length == 0)
		return nullptr;

	node* root = new node;
	root->key = list[0];
	root->parent = nullptr;
	root->left = nullptr;
	root->right = nullptr;

	for(int i = 1; i < length; ++i)
		insert(root, list[i]);

	return root;
}

void traverse(node* current, int* list) {
	if(current == nullptr)
		return;

	traverse(current->left, list);
	list[i++] = current->key;
	traverse(current->right, list);
}

void bst_sort(int* list, int length) {
	// Build BST
	node* root = build_binary_tree(list, length);

	// Traverse the BST in order
	i = 0;
	traverse(root, list);
}

int main() {
	int length;

	cout << "Enter the number of values you wish to input: ";
	cin >> length;
	int list[length];

	for(int i = 0; i < length; ++i) {
		cout << "Number " << (i + 1) << ": ";
		cin >> list[i];
	}
	bst_sort(list, length);

	for(int i = 0; i < length; ++i)
		cout << list[i] << " ";
	cout << "\n";

	return 0;
}